dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -183,61 +179,111 @@ public void setInterceptorChain(InterceptorChain ic) {
 
     @Override
     public Object put(String key, Object value) {
-        if (contextCache != null) {
-            contextCache.put(key, value);
-        }
         return super.put(key, value);
     }
 
     @Override
     public Object getContextualProperty(String key) {
-        if (contextCache == null) {
-            calcContextCache();
+        Object o = super.get(key);
+        if (o != null) {
+            return o;
         }
-        return contextCache.get(key);
+        return getFromExchange(key);
```

Comment:
This is a behavior change that should be validated with the JAX-RS team.  Before this change on first call to getContextualProperty causes a snapshot of the properties in the Bus, Service and Endpoint to be created.  If properties in the Bus, Service and Endpoint are updated, added, or removed, the contextCache would not know about those modifications.  Now with the getFromExchange method, it will see any changes made since the first getContextualProperty call without the need for the resetContextCache being called.